src/server/cluster/incoming_slot_migration.cc

```diff
@@ -85,15 +101,23 @@ IncomingSlotMigration::IncomingSlotMigration(string source_id, Service* se, Slot
 }
 
 IncomingSlotMigration::~IncomingSlotMigration() {
-  sync_fb_.JoinIfNeeded();
 }
 
 void IncomingSlotMigration::Join() {
   bc_->Wait();
   state_ = MigrationState::C_FINISHED;
 }
 
-void IncomingSlotMigration::StartFlow(uint32_t shard, io::Source* source) {
+void IncomingSlotMigration::Cancel() {
+  cntx_.Cancel();
+  for (auto& flow : shard_flows_) {
```

Comment:
flows are created in different threads, I think you need AwaitFiberOnAll(cencel_cb);